mozjs-sys/src/jsimpls.rs

```diff
@@ -386,30 +387,66 @@ impl JSNativeWrapper {
 impl<T> JS::Rooted<T> {
     pub fn new_unrooted() -> JS::Rooted<T> {
         JS::Rooted {
-            stack: ptr::null_mut(),
-            prev: ptr::null_mut(),
+            base: RootedBase {
+                stack: ptr::null_mut(),
+                prev: ptr::null_mut(),
+            },
+            ptr: unsafe { std::mem::zeroed() },
+        }
+    }
+
+    pub unsafe fn add_to_root_stack(&mut self, cx: *mut JSContext)
+    where
+        T: RootKind,
+    {
+        let kind = T::rootKind();
+        // Any stack root that uses the Traceable kind needs to use RootedTraceable instead.
+        // Using Traceable with Rooted<T> leads to crashes during GCs.
+        assert_ne!(kind, JS::RootKind::Traceable);
+        self.base.add_to_root_stack(cx, kind)
+    }
```

Comment:
I am not really thrilled with this runtime check, I think I managed to put this in compiletime by merging Rooted and RootedTraceable back into one type: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=cc1af6037a0b3c6373e12116585d9c28

Reply:
Oooh, that's really interesting. I'll see if I can integrate that!

Reply:
Grumble. The data representation isn't quite correct when view from the C++ code in lldb, and I can't figure out why yet. The vtable pointer is being treated as the value of the stack pointer.

Reply:
do you have proper repr attribute everywhere (even on vtable type)?

Reply:
https://github.com/servo/mozjs/compare/main...jdm:mozjs:propdesc2?expand=1

Reply:
Maybe try without TraceableTrace hop:
```
mod prop_desc {
    const vftable: RootedVFTable = RootedVFTable {
        padding: RootedVFTable::PADDING,
        trace,
    };

    unsafe extern "C" fn trace(this: *mut c_void, trc: *mut JSTracer, _name: *const c_char) {
        let rooted = this as *mut RootedTraceable<Self>;
        let rooted = rooted.as_mut().unwrap();
        CallPropertyDescriptorTracer(trc, &mut rooted.ptr);
    }
}

impl RootKind for JS::PropertyDescriptor {
    type Vtable = *const RootedVFTable;
    const VTABLE: Self::Vtable = &prop_desc::vftable;
    const KIND: JS::RootKind = JS::RootKind::Traceable;
}
```

Reply:
> The vtable pointer is being treated as the value of the stack pointer.

This could also be coming from using & instead of addr_of!(), another potential problem could be using `const` instead of `static` (I think consts are most of the time inlined).

Reply:
Sadly none of those ideas made a difference.

Reply:
Oh, I just realized what the difference in code is—in the first branch, I moved add_to_root_stack/remove_from_root_stack to RootedBase, which is after the vtable field, but it used self as `Rooted<T>`. That seems extremely sketchy, but it apparently worked :<

Reply:
Yeah, https://github.com/servo/mozjs/commit/712f584e0b94674fdb9a157a1b060c39a87058af made the test pass. Not sure how I feel about that!

Reply:
We do not need RootedBase if we only have one rooted type.

Reply:
Ah, `TraceExactStackRootTraceableList` expects `StackRootedBase*` on its trace list and this is then casted to `*StackRootedTraceableBase`, so if we pass ptr to Rooted (that already has vtable inside) we get dummy data for vtable: https://searchfox.org/mozilla-central/rev/c0b40fc662d376ecdee66b5d7489f4be63ef88bc/js/src/gc/RootMarking.cpp#56

Reply:
```patch
diff --git a/mozjs-sys/src/jsimpls.rs b/mozjs-sys/src/jsimpls.rs
index 3a9efc5f0..ca631cb37 100644
--- a/mozjs-sys/src/jsimpls.rs
+++ b/mozjs-sys/src/jsimpls.rs
@@ -412,11 +412,11 @@ impl<T: RootKind> JS::Rooted<T> {
         self.base.stack = stack;
         self.base.prev = *stack;
 
-        *stack = self as *mut _ as usize as _;
+        *stack = ::std::ptr::addr_of_mut!(self.base) as usize as _;
     }
 
     pub unsafe fn remove_from_root_stack(&mut self) {
-        assert!(*self.base.stack == self as *mut _ as usize as _);
+        assert!(*self.base.stack == ::std::ptr::addr_of_mut!(self.base) as usize as _);
         *self.base.stack = self.base.prev;
     }
 }
```
on top of https://github.com/servo/mozjs/commit/712f584e0b94674fdb9a157a1b060c39a87058af also works, but I like RootedBase solution better, but we should write comment in code why is that also inline markings.